### PR TITLE
Adds some simple methods to Timer.

### PR DESCRIPTION
**Pull request by thalesfragoso**
Very simple methods but potentially useful, especially the one to clear the update interrupt flag.
I was having trouble using the timer with interrupts, because the Timer takes ownership of the registers and the `tim` field is private.
Sorry if I got something wrong, I am kinda new to rust.

### CODE REVIEW

**Comment by therealprof** [comment]
Looks okay to me but there's no need to release `Clocks` since it's a copyable resource.

**Comment by thalesfragoso** [comment]
> Looks okay to me but there's no need to release `Clocks` since it's a copyable resource.

It makes sense, I did not realize that before. I have changed that bit in the new commit.

File: src/timer.rs [hunk 1]
```diff
@@ -72,6 +72,17 @@ impl Timer<SYST> {
             Event::Update => self.tim.disable_interrupt(),
         }
     }
+
+    /// Stops the timer
+    pub fn stop(&mut self) {
+        self.tim.disable_counter();
+    }
+
+    /// Releases the SYST and clocks
```
Comment (therealprof) [inline src/timer.rs]:
```suggestion
    /// Releases the SYST
```

File: src/timer.rs [hunk 2]
```diff
@@ -133,6 +144,22 @@ macro_rules! hal {
...
+                }
+
+                /// Clears Update Interrupt Flag
+                pub fn clear_uif(&mut self) {
+                    self.tim.sr.modify(|_, w| w.uif().clear_bit());
+                }
+
+                /// Releases the TIM Peripheral and clocks
+                pub fn release(self) -> $TIMX {
+                    self.tim.cr1.modify(|_, w| w.cen().clear_bit());
```
Comment (therealprof) [inline src/timer.rs]: Can't we simply call `stop()` from above?

**Comment by TheZoq2** [comment]
Looks good to me

**Comment by thalesfragoso** [comment]
I updated the Changelog, don't know if I over-explained it, feel free to modify.

**Comment by TheZoq2** [comment]
Looks good to me, though I did think of another improvement which would be nice. Could you replace the already existing calls to `uif.clear_bit()` with the new `clear_uif` function?

Also, I would argue that calling that function `clear_update_interrupt` might be more intuitive than `clear_uif`

**Comment by thalesfragoso** [comment]
> Looks good to me, though I did think of another improvement which would be nice. Could you replace the already existing calls to `uif.clear_bit()` with the new `clear_uif` function?
> 
> Also, I would argue that calling that function `clear_update_interrupt` might be more intuitive than `clear_uif`

How about `clear_update_interrupt_flag` ? Flag is a very common word and would be the first one I would search for in a function to do this procedure.

**Comment by TheZoq2** [comment]
Yea, that is probably even better!

File: CHANGELOG.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - Add feature for using STM32F101 chip
 - Add gpio pins corresponding to LQFP-100 package
 - Implement `core::fmt::Write` for `serial::Tx`
+- Add methods `stop`, `release` and `clear_uif` to `Timer` (`clear_uif` only on `Timer<TIMX>`)
```
Comment (TeXitoi) [inline CHANGELOG.md]: unsync

File: src/timer.rs [hunk 4]
```diff
@@ -72,6 +72,17 @@ impl Timer<SYST> {
...
     }
+
+    /// Stops the timer
+    pub fn stop(&mut self) {
+        self.tim.disable_counter();
+    }
+
+    /// Releases the SYST
+    pub fn release(mut self) -> SYST {
+        self.tim.disable_counter();
```
Comment (TeXitoi) [inline src/timer.rs]: minor: better to `self.stop()` as any bug on stop must be propagated here.

**Comment by thalesfragoso** [comment]
I will modify the code to use the new methods as proposed and I think I will write an example for the Timer interrupt with rtfm and add it to the PR.
I will probably do it over the weekend.

**Comment by thalesfragoso** [comment]
I ended up getting some spare time today, so there you go. Sorry for the bunch of commits, but I guess you gonna squash them all anyways. 
I tested the example on a bluepill and it worked as expected.

**Comment by TheZoq2** [comment]
Thanks, looks mostly good to me!

@therealprof What's our policy on adding dependencies for examples, #2 didn't really come to a conclusion. I believe that rtfm was originally removed because it didn't compile on stable, but that's solved now.

**Comment by thalesfragoso** [comment]
I kinda understand the dependencies for examples thing, but in the present scenario I would think that it is somewhat bad to push newcomers to implement interrupts (with handlers and stuff) without rtfm.

**Comment by therealprof** [comment]
@TheZoq2 I'm generally not too hot on having examples in this crate (apart from in the docs). Ideally we should have BSP crates which go nuts on examples. I do have a a basic https://github.com/therealprof/nucleo-f103rb crate, if people are interested I could make this free for all, or just create a new blue pill crate.

However for now it's okay for me to just add back the RTFM dependency.

**Comment by therealprof** [comment]
@thalesfragoso The problem is that a universal HAL crate is somewhat tricky to use anyway, especially for beginners; so instead of dragging newcomers into examples in here we really should demonstrate how to get properly started with your own projects by having a proper BSP crate.

File: examples/timer-interrupt-rtfm.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+        // Count used to change the timer update frequency
+        static mut COUNT: u8 = 0;
+
+        if *resources.LED_STATE {
+            // Uses resourcers managed by rtfm to turn led off (on bluepill)
+            resources.LED.set_high();
+            *resources.LED_STATE = false;
+        } else {
+            resources.LED.set_low();
```
Comment (TeXitoi) [inline examples/timer-interrupt-rtfm.rs]: You can call `resources.LED.toggle();` and remove LED_STATE

File: examples/timer-interrupt-rtfm.rs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            TIMER_HANDLER: timer,
+        }
+    }
+
+    #[idle]
+    fn idle() -> ! {
+
+        loop {
+            // Waits for interrupt
+            wfi();
```
Comment (TeXitoi) [inline examples/timer-interrupt-rtfm.rs]: You can just not define an `idle()` function.

**Comment by TheZoq2** [comment]
Yea, eventually we should probably take care of this example stuff, and a BSP with examples does sound like a good idea. I do agree that having a RTFM example for interrupts is much better than one without RTFM :)

I guess all that's left to do then is to merge this, thanks for the help! @thalesfragoso

File: examples/timer-interrupt-rtfm.rs (hunk 5 above)
Comment (thalesfragoso) [reply examples/timer-interrupt-rtfm.rs]: Does PC13<Output<MODE>> have toggle ? I remember looking at the docs when I was writing other code and could not find it.

File: examples/timer-interrupt-rtfm.rs (hunk 6 above)
Comment (thalesfragoso) [reply examples/timer-interrupt-rtfm.rs]: Does it automatically call wfi() ?

File: examples/timer-interrupt-rtfm.rs (hunk 5 above)
Comment (TeXitoi) [reply examples/timer-interrupt-rtfm.rs]: I've used it in the past https://github.com/TeXitoi/blue-pill-quickstart/blob/master/src/main.rs#L41

File: examples/timer-interrupt-rtfm.rs (hunk 6 above)
Comment (TeXitoi) [reply examples/timer-interrupt-rtfm.rs]: https://japaric.github.io/cortex-m-rtfm/book/en/by-example/app.html#idle

> When no idle function is declared, the runtime sets the SLEEPONEXIT bit and then sends the microcontroller to sleep after running init.

**Pull request merged**